pkg/api/client/client.go

```diff
@@ -0,0 +1,183 @@
+
+	name string
+
+	requestLogging bool
+}
+
+type Option func(*DefaultClient)
+
+func WithRequestLogging(b bool) Option {
+	return func(c *DefaultClient) {
+		c.requestLogging = true
+	}
+}
+
+func WithRequestTimeout(timeout time.Duration) Option {
+	return func(c *DefaultClient) {
+		if timeout > 0 {
+			c.timeout = timeout
+		}
+	}
+}
+
+func WithAPIKey(key string) Option {
+	return func(c *DefaultClient) {
+		if key != "" {
+			c.key = key
+		}
+	}
+}
+
+func WithScheme(scheme string) Option {
+	return func(c *DefaultClient) {
+		if scheme != "" {
+			c.scheme = scheme
+		}
+	}
+}
+
+func WithName(name string) Option {
+	return func(c *DefaultClient) {
+		if name != "" {
+			c.name = fmt.Sprintf("%s/%s", name, version.Short())
+		}
+	}
+}
+
+const (
+	defaultRequestTimeout = 30 * time.Second
+	defaultClientName     = "default-client"
+)
+
+// NewDefault creates a new default client that can be used for all calls to goProbe APIs
+func NewDefault(addr string, opts ...Option) *DefaultClient {
+	c := &DefaultClient{
+		client:   http.DefaultClient,
+		scheme:   "http",
+		hostAddr: addr,
+		timeout:  defaultRequestTimeout,
+		name:     defaultClientName,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	c.client = &http.Client{
+		Transport: &transport{
+			rt: otelhttp.NewTransport(
+				http.DefaultTransport,
+			),
+			requestLogging: c.requestLogging,
+			clientName:     c.name,
+		},
+	}
+	return c
+}
+
+func (c *DefaultClient) Client() *http.Client {
+	return c.client
+}
+
+type transport struct {
+	rt             http.RoundTripper
+	requestLogging bool
+	clientName     string
+}
+
+func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
+	start := time.Now()
+
+	r.Header.Set("User-Agent", t.clientName)
+
+	// make request
+	resp, err := t.rt.RoundTrip(r)
+	duration := time.Since(start)
+
+	if t.requestLogging {
+		logger := logging.WithContext(r.Context()).With("req", slog.GroupValue(
+			slog.String("method", r.Method),
+			slog.String("url", r.URL.String()),
+			slog.String("user-agent", r.UserAgent()),
+			slog.Duration("duration", duration),
+		))
+		// log trace ID if it is present
+		sc := trace.SpanContextFromContext(r.Context())
+		if sc.HasTraceID() {
+			logger = logger.With(slog.String("traceID", sc.TraceID().String()))
+		}
+
+		if err == nil {
+			if resp != nil {
+				logger = logger.With("resp", slog.GroupValue(
+					slog.Int("status_code", resp.StatusCode),
+				))
+				switch {
+				case 200 <= resp.StatusCode && resp.StatusCode < 300:
+					logger.Info("completed request")
+				case 300 <= resp.StatusCode && resp.StatusCode < 400:
+					logger.Info("further action needed to complete request")
+				default:
+					logger.Error("server error returned")
+				}
+			} else {
+				logger.Error("empty response")
+			}
+		} else {
+			logger.Errorf("failed to send request: %v", err)
+		}
+	}
+	return resp, err
+}
+
+func (c *DefaultClient) Modify(ctx context.Context, req *httpc.Request) *httpc.Request {
```

Comment:
It can, but I want to add the key for every client request. Right now, only `Query` is implemented. Once the client is richer, there will be other methods.

And I don't want to explicitly set the key in each of those methods and then forgetting it in one place.

Hence the wrapper method from the client.

Reply:
Mh, I'd like to better understand that, maybe we can have a chat about that some other time to see if the package can be made better. Ideally, I want the flow to be as convenient as possible and in one place. The whole chaining approach is weaker once you have to fiddle with individual aspects outside of its scope...